### PR TITLE
Integrate gRPC changes to fix #6116

### LINKED ISSUE #6116: grpc RecvTensor is slow
I made benchmark tests for distributed setup with loopback network, profiling it and found there is excessive memory copying in the client side of RecvTensor call, which is actually one of the bottleneck.

Here is the code, which mainly stolen from @yaroslavvb [here](https://gist.github.com/yaroslavvb/1124bb02a9fd4abce3d86caf2f950cb2),
```python
  with tf.device(device1):                                                      
    params = tf.get_variable("params", shape=[params_size], dtype=dtype,        
                             initializer=tf.zeros_initializer)                  
  with tf.device(device2):                                                      
    # constant node gets placed on device1 because of simple_placer             
    #    update = tf.constant(1, shape=[params_size], dtype=dtype)              
    update = tf.get_variable("update", shape=[params_size], dtype=dtype,        
                             initializer=tf.ones_initializer())                 
    add_op = params.assign(update)
```

Here is the profiling result (google perftools) with tensor size 100MB (one fact is, the throughput will degrade with the increasing of tensor size):
* [device1 worker profiling report](https://github.com/tensorflow/tensorflow/files/633183/device1-prof.pdf)
* [device2 worker profiling report](https://github.com/tensorflow/tensorflow/files/633184/device2-prof.pdf)

From the result, the sending side (device2) look fine, but the receiving side (device1, the grpc client) consumes too many CPU cycles for the data transfer.

By the way, I made rough stats for this [memmove](https://github.com/grpc/grpc/blob/d7ff4ff40071d2b486a052183e3e9f9382afb745/src/core/lib/support/slice_buffer.c#L278) call. For one round of 100MB tensor assignment, there are roughly 2GB data moved (actually, including the copy inside memmove, it should be 4GB copied with a naive memmove), which is 20+ times RAM bandwidth amplification (the result is an average of 100 round run, which may not precise but the scale should be ok).

### PR DESCRIPTION
**Pull request by llhe**
The gRPC won't build for some reasons, first make a hackish fix just for the review.

### CODE REVIEW

**Comment by llhe** [comment]
This is not complete, do not merge.

**Comment by llhe** [comment]
rpcbench_test results:
* without the grpc fix
```
BM_ShardedProgram/1/1         125990       5604	3 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/3         515428       1000	5 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/5         673779        987	7 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/15       1366261        456	17 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/60       5010130        100	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/1        873925        837	17 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/3       3341377        212	47 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/5       5391330        100	77 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/1       1588803        437	32 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/2       6168890        100	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/3      11095890        100	92 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/5      20620620        100	152 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/1       2921452        230	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/3      44074260        100	182 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/5      83405230        100	302 nodes; Multi device; tensor bytes/send: 8
BM_RPC/30/2                  6221740        100	62 nodes; Multi device; tensor bytes/send: 8
BM_RPC/30/1000               7289430        100	62 nodes; Multi device; tensor bytes/send: 4000
BM_RPC/30/100000            98748280        100	62 nodes; Multi device; tensor bytes/send: 400000
BM_RPC/1/100000000        4321044180        100	4 nodes; Multi device; tensor bytes/send: 400000000
BM_SingleDevice/1/1           106932       6345	3 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/30/2          102905       5610	62 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/60/5          150717       5097	302 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/4/10000      9268440        100	40002 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/1/1000000  448853650        100	1000002 nodes; Single device; tensor bytes/send: 8
```
* with the grpc fix
```
BM_ShardedProgram/1/1         105292       6073	3 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/3         823504       1000	5 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/5         698704        961	7 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/15       1412678        460	17 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/1/60       5179623        146	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/1        960009        738	17 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/3       3826291        182	47 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/15/5       6762430        100	77 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/1       1621082        404	32 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/2       7305350        100	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/3      16084080        100	92 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/30/5      26449080        100	152 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/1       3076433        208	62 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/3      55058800        100	182 nodes; Multi device; tensor bytes/send: 8
BM_ShardedProgram/60/5     111056120        100	302 nodes; Multi device; tensor bytes/send: 8
BM_RPC/30/2                  6928160        100	62 nodes; Multi device; tensor bytes/send: 8
BM_RPC/30/1000               8062080        100	62 nodes; Multi device; tensor bytes/send: 4000
BM_RPC/30/100000            95994550        100	62 nodes; Multi device; tensor bytes/send: 400000
BM_RPC/1/100000000        1332521740        100	4 nodes; Multi device; tensor bytes/send: 400000000
BM_SingleDevice/1/1           112106       5918	3 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/30/2          134409       4993	62 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/60/5          154851       4168	302 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/4/10000     13725570        100	40002 nodes; Single device; tensor bytes/send: 8
BM_SingleDevice/1/1000000  446881410        100	1000002 nodes; Single device; tensor bytes/send: 8
```
Actually this benchmark test looks quite noisy and unstable, just pay attention of this case (newly added in this patch):
```
BM_RPC/1/100000000        4321044180        100	4 nodes; Multi device; tensor bytes/send: 400000000
```
```
BM_RPC/1/100000000        1332521740        100	4 nodes; Multi device; tensor bytes/send: 400000000
```

File: tensorflow/workspace.bzl
Comment (jhseu) [inline tensorflow/workspace.bzl]: Is this rename necessary? There's nothing referring to `//external:protobuf`

**Comment by jhseu** [comment]
Changes look good to me (besides the one comment). Let me know when it's ready for merging.

**Comment by llhe** [comment]
@jhseu "//external:protobuf" is [referenced here](https://github.com/llhe/tensorflow/blob/19bf520f381b50086aadb2eb41eabfeca55d0e3e/third_party/grpc.BUILD#L1276).

File: tensorflow/workspace.bzl
Comment (llhe) [inline tensorflow/workspace.bzl]: "//external:protobuf" is [referenced here](https://github.com/llhe/tensorflow/blob/19bf520f381b50086aadb2eb41eabfeca55d0e3e/third_party/grpc.BUILD#L1276).

**Comment by llhe** [comment]
Merged with the master for the protobuf update.
@jhseu I made a quick and dirty fix over grpc.BUILD, but I think it can be merged now. We can update it when grpc side fix it's build file.

**Comment by jhseu** [comment]
(Testing first)
Jenkins, test this please

File: tensorflow/tools/lib_package/BUILD
Comment (jhseu) [inline tensorflow/tools/lib_package/BUILD]: We need to include this license for the sanity checks to pass. You can run them yourself under `tensorflow/tools/ci_build/ci_sanity.sh`.

I just sent a change to gRPC so that the license file is exported so it can be used here:
https://github.com/grpc/grpc/pull/9711

Comment (jhseu) [inline tensorflow/tools/lib_package/BUILD]: That gRPC change was committed. Can you:
- Update to the latest gRPC commit.
- Rebase with tensorflow master.
- And add the license back here

Thanks!

File: tensorflow/workspace.bzl
Comment (jhseu) [inline tensorflow/workspace.bzl]: Uncomment?

Comment (llhe) [inline tensorflow/workspace.bzl]: It triggers 404 now, I don't know the magic how to fill the mirror. It can be uncommitted when the mirror has this file.

File: tensorflow/tools/lib_package/BUILD
Comment (llhe) [inline tensorflow/tools/lib_package/BUILD]: OK

**Comment by llhe** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

File: tensorflow/tools/lib_package/BUILD
Comment (jhseu) [inline tensorflow/tools/lib_package/BUILD]: Please restore this line

File: tensorflow/tools/lib_package/BUILD
Comment (jhseu) [inline tensorflow/tools/lib_package/BUILD]: And this one

Comment (jhseu) [inline tensorflow/tools/lib_package/BUILD]: (It probably fails locally with them because of your ./configure options)

Comment (llhe) [inline tensorflow/tools/lib_package/BUILD]: oh, my fault, I see

Comment (llhe) [inline tensorflow/tools/lib_package/BUILD]: restored

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
@mrry this fails the windows build, but passes everything else. Presumably https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/patches/grpc/CMakeLists.txt needs to be updated.

Not planning to learn cmake at the moment :) Derek, how do you want to proceed?

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by mrry** [comment]
There's not too much to change:

1. Update [this line](https://github.com/tensorflow/tensorflow/blob/89059e6f0b2788b624e744afc21ba2472523a250/tensorflow/contrib/cmake/external/grpc.cmake#L6) to refer to the new commit for gRPC.
2. Copy each of the file lists from the corresponding `cc_library` in `grpc.BUILD` to the `add_library()` blocks in [`grpc/CMakeLists.txt`](https://github.com/tensorflow/tensorflow/blob/89059e6f0b2788b624e744afc21ba2472523a250/tensorflow/contrib/cmake/patches/grpc/CMakeLists.txt).

**Comment by jhseu** [comment]
Ah, ok, I'll make an attempt later today.

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
@mrry So the latest gRPC strictly requires SSL through the census code. It might be possible to unravel by patching the code, but I'm not sure it's worth it.

I updated the CMake files to make boringssl required, and the build works (on Linux at least), but adds additional build-time dependencies for Windows. You can see that list here:
https://github.com/google/boringssl/blob/master/BUILDING.md

Since it's rare for people to build for Windows from scratch (they'd more likely use the package we build), should we just install those dependencies on our Windows machines and require boringssl from now on?

**Comment by mrry** [comment]
Hmm, the build of boringssl on Windows is going to be pretty tedious. I don't know much about Census, but it looks like the only SSL reference in its code is [here](https://github.com/grpc/grpc/blob/36ffc6e0f73a1b13a79bf120b7a2517136b8e5d5/src/core/ext/census/tracing.c#L39), in a file with a bunch of no-op implementations that doesn't appear to use any of the included headers.

To avoid changing the test infrastructure (and making the build more complicated for a library we don't end up using), could we do one of the following:

* Provide an empty stub for `<openssl/rand.h>`, like we do for [`<unistd.h>`](https://github.com/tensorflow/tensorflow/blob/43c71a03380d8de18202cc399563814b2f438cd2/tensorflow/contrib/cmake/patches/gif/unistd.h) when building giflib?
* Avoid building the Census source altogether, since IIUC it isn't referenced from any other part of TensorFlow or gRPC?

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Review by jhseu — APPROVED** [review]

**Comment by jhseu** [comment]
@mrry Thanks! The empty file worked.

**Comment by jhseu** [comment]
Note: currently blocked on a large RPCs bug for gRPC on Windows. See https://github.com/grpc/grpc/pull/9826

**Comment by drpngx** [comment]
@llhe please resolve changes when you get a chance, thanks!

**Comment by llhe** [comment]
Jenkins, test this please

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
License check failed. Please add the copyright notice to all files. Thanks!

**Comment by jhseu** [comment]
Note that we're blocked on a few gRPC issues that @vjpai is working on. I'll update this pull request when it's ready.

**Comment by vjpai** [comment]
Note that the bazel issues in grpc should be taken care of by the recently merged grpc/grpc#10639

**Comment by martinwicke** [comment]
@jhseu is this waiting for another grpc fix os is this it?

**Comment by jhseu** [comment]
@martinwicke We're waiting on one more fix.

**Comment by vrv** [comment]
@jhseu is there an ETA on the fix?  Is there a PR already open for it?

**Comment by jhseu** [comment]
@vrv We're waiting on a fix for https://github.com/grpc/grpc/issues/10161

I'll try syncing this pull request to the latest gRPC commit later today, though.

**Comment by jhseu** [comment]
(Note to self) Windows testing happening here:
https://github.com/grpc/grpc/pull/10716

**Comment by vrv** [comment]
@jhseu what's the current status?

**Comment by vjpai** [comment]
@ctiller will be picking up the windows debugging issue.

**Comment by vrv** [comment]
(in other words, we're still waiting on some fix to gRPC?)

**Comment by jhseu** [comment]
@vrv Yep

**Comment by rmlarsen** [comment]
@llhe could you rebase to fix conflicts? 
What is the status here? Are we still waiting for grpc fixes? i'll mark this as stalled for now.

**Comment by llhe** [comment]
@rmlarsen  We are still waiting changes from gRPC.

**Comment by martinwicke** [comment]
Any news? This would have to be rebased. I'm tempted to close this until the gRPC changes are ready.

**Comment by byronyi** [comment]
I think this PR is important enough (3x faster large tensor copies) to remain tracked until it is merged. By the way why does it take so long in gRPC side? It is literally been 6 months and it is supposed to be merged into v1.0 [initially](https://github.com/tensorflow/tensorflow/issues/6116#issuecomment-269716302).

**Comment by drpngx** [comment]
Periodic ping to check status.

**Comment by tfboyd** [comment]
Update as of 22-JUN.  The gRPC team is assigning someone someone to clear this issue.  This issue was escalated and is being watch closely.  If updates are not seen feel free to ping this thread/issue.

**Comment by fangy** [comment]
ping to check status

**Comment by promiseofcake** [comment]
> Update as of 22-JUN. The gRPC team is assigning someone someone to clear this issue.
Do we know where this stands, or is there a GRPC issue we can follow?

**Comment by martinwicke** [comment]
@promiseofcake we're waiting on https://github.com/grpc/grpc/issues/10161

**Comment by promiseofcake** [comment]
Will this: https://github.com/tensorflow/tensorflow/pull/11768, render this change moot?

**Comment by jhseu** [comment]
@promiseofcake Yep, closing this pull request in favor of the new change. It was easier to redo from scratch than to bring this change up-to-date.

**Comment by luchensk** [comment]
@jhseu Just to make sure, has the fix [#6116](https://github.com/tensorflow/tensorflow/issues/6116) been checked into master branch of tensorflow? Thanks.

**Comment by jhseu** [comment]
Yep

**Pull request closed without merging**
